## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [3]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [4]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [5]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [6]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [7]:
# What three item categories had the most entries?
top_categories= df['category'].value_counts().head(3)
print(top_categories)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [8]:
# For the category with the most entries, which subcategory had the most entries?
top_category = df['category'].value_counts().idxmax()
filtered_df = df[df['category'] == top_category]
top_subcategory = filtered_df['subcategory'].value_counts().idxmax()
print(f"Top Category: {top_category}")
print(f"Top Subcategory within '{top_category}': {top_subcategory}")

Top Category: consumables
Top Subcategory within 'consumables': bathroom supplies


In [9]:
# Which five clients had the most entries in the data?
top_clients= df['client_id'].value_counts().head(5)
print(top_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [10]:
# Store the client ids of those top 5 clients in a list.
top_clients= df['client_id'].value_counts().head(5).index.tolist()
print(top_clients)

[33615, 66037, 46820, 38378, 24741]


In [11]:
# How many total units (the qty column) did the client with the most entries order order?
top_client = df['client_id'].value_counts().idxmax()
total_qty = df[df['client_id'] == top_client]['qty'].sum()
print(f"Client ID with the most entries: {top_client}")
print(f"Total quantity purchased by {top_client}: {total_qty}")

Client ID with the most entries: 33615
Total quantity purchased by 33615: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [12]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df.head(5)

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [13]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda weight: 7 * weight if weight > 50 else 10 * weight)
df.head(5)


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  total_weight  \
0     762.71         7.50  105            1  115164.00        787.50   
1      15.09         1.49   21            0     523.95         31.29   
2       7.86         1.68   39            6     527.28         65.52   
3      24.85         1.23   29            3    1056.18         35.67   
4     108.17        46.43   20            1    3902.00        928.60   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [15]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = df['unit_price'] * df['qty']
df['tax'] = df['total_price'] * 0.0925
df['total_sales'] = df['total_price'] + df['tax'] + df['shipping_price']
df.head(5)

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  total_weight  shipping_price  total_price           tax  \
0  115164.00        787.50         5512.50    115164.00  10652.670000   
1     523.95         31.29          312.90       523.95     48.465375   
2     527.28         65.52          458.64       527.28     48.773400   
3    1056.18         35.67          356.70      1056.18     97.696650   
4    3902.00        928.60         6500.20      3902.00    360.935000   

     total_sales  
0  131329.170000  
1     885.315375  
2    1034.693400  
3    1510.576650  
4   10763.135000  

[5 rows x 24 columns]

In [16]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['total_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df.head(5)


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number   subtotal  \
0          17        2023  ...        7.50  105           1  115164.00   
1          20        2023  ...        1.49   21           0     523.95   
2           4        2023  ...        1.68   39           6     527.28   
3          17        2023  ...        1.23   29           3    1056.18   
4          21        2023  ...       46.43   20           1    3902.00   

   total_weight  shipping_price  total_price           tax    total_sales  \
0        787.50         5512.50    115164.00  10652.670000  131329.170000   
1         31.29          312.90       523.95     48.465375     885.315375   
2         65.52          458.64       527.28     48.773400    1034.693400   
3         35.67          356.70      1056.18     97.696650    1510.576650   
4        928.60         6500.20      3902.00    360.935000   10763.135000   

   total_cost  
0    85597.05  
1      629.79  
2      765.18  
3     1077.35  
4     8663.60  

[5 rows x 25 columns]

In [17]:
# Create a column for the profit of each line using line cost and line price
df['total_profit'] = df['total_price'] - df['total_cost']
df.head(5)


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...  qty line_number   subtotal  total_weight  \
0          17        2023  ...  105           1  115164.00        787.50   
1          20        2023  ...   21           0     523.95         31.29   
2           4        2023  ...   39           6     527.28         65.52   
3          17        2023  ...   29           3    1056.18         35.67   
4          21        2023  ...   20           1    3902.00        928.60   

   shipping_price  total_price           tax    total_sales  total_cost  \
0         5512.50    115164.00  10652.670000  131329.170000    85597.05   
1          312.90       523.95     48.465375     885.315375      629.79   
2          458.64       527.28     48.773400    1034.693400      765.18   
3          356.70      1056.18     97.696650    1510.576650     1077.35   
4         6500.20      3902.00    360.935000   10763.135000     8663.60   

   total_profit  
0      29566.95  
1       -105.84  
2       -237.90  
3        -21.17  
4      -4761.60  

[5 rows x 26 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [18]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]
summary_df = df[df['order_id'].isin(order_ids)].groupby('order_id')['total_sales'].sum().reset_index()
summary_df.columns = ['order_id', 'sum_total_sales']
print(summary_df)

   order_id  sum_total_sales
0   2173913    158762.830300
1   2742071    151695.028025
2   6128929    908947.334975


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [19]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_5_clients = df.groupby('client_id')['qty'].sum().nlargest(5).index
summary_df = df[df['client_id'].isin(top_5_clients)].groupby('client_id')['total_sales'].sum().reset_index()
summary_df.columns = ['client_id', 'sum_total_sales']
print(summary_df)

   client_id  sum_total_sales
0      30143     3.249802e+07
1      77354     6.973979e+07
2      78965     6.554297e+08
3      85931     2.364807e+08
4      86721     1.753152e+08


In [20]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = (
    df[df['client_id'].isin(top_5_clients)]
    .groupby('client_id')
    .agg(
        total_units_purchased=('qty', 'sum'),
        total_shipping_price=('shipping_price', 'sum'),
        total_price=('total_sales', 'sum'),
        total_profit=('total_profit', 'sum')
    )
    .reset_index()
)
print(summary_df)

   client_id  total_units_purchased  total_shipping_price   total_price  \
0      30143                 846069            8981905.24  3.249802e+07   
1      77354                 506382           23814685.86  6.973979e+07   
2      78965                4033302            5340756.88  6.554297e+08   
3      85931                 644423           10122549.19  2.364807e+08   
4      86721                1010426           64335727.72  1.753152e+08   

   total_profit  
0 -5.534128e+05  
1 -1.747368e+07  
2  1.864762e+08  
3  7.906991e+07  
4 -3.670852e+07  


In [21]:
# Define a function that converts a dollar amount to millions.
def convert_to_millions(value):
    return round(value / 1_000_000, 2)

# Apply the currency_format_millions function to only the money columns. 
# Rename the columns to reflect the change in the money format. 
columns_to_convert = ['total_price', 'total_cost', 'total_sales', 'shipping_price', 'total_profit']
for column in columns_to_convert:
    client_totals[f"{column} (in millions)"] = client_totals[column].apply(convert_to_millions)
client_totals.drop(columns=columns_to_convert, inplace=True)
print("\nUpdated DataFrame with columns in millions:")
print(client_totals)

#I can't figure out why this one thing doesn't run.


NameError: name 'client_totals' is not defined

In [23]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_df = df.sort_values(by='total_profit', ascending=False)
print(sorted_df)

         first       last                                       job  \
21801    Julie   Anderson                       Exhibition designer   
15112   Dennis    Hawkins       Social research officer, government   
4063    Kendra    Garrett                    Pension scheme manager   
23929  Michele   Mcdonald     Emergency planning/management officer   
11740   Lauren      Jones                                 Architect   
...        ...        ...                                       ...   
40161  Charles  Mccormick  Conservation officer, historic buildings   
29737  William     Savage                         Minerals surveyor   
50048   Amanda    Burnett                           Records manager   
12597   Jordan      Price                  Local government officer   
25972      Amy    Stewart                 Psychologist, educational   

                       phone                              email  client_id  \
21801             5933099462      anderson_better_4704@bell.biz      